Add EVP/KDF API

davidmakepeace opened the pull request:
This PR adds a new EVP/KDF API that collects all KDFs and PRFs under the one API.

`#include <openssl/kdf.h>
`

Currently some KDFs are available directly (PBKDF2 and scrypt) while others are available through the PKEY API (scrypt, TLS1 PRF and HKDF) even though the PKEY API was originally intended for Public Key algorithms. Note that the low level KDF functions (PKCS5_PBKDF2_HMAC and EVP_PBE_scrypt) are defined in openssl/evp.h even though EVP is the high level interface.

The new API is modeled after the current KDF support under the PKEY API in order to minimize the work required to update applications to use the new API. Support is included for PBKDF2, scrypt, TLS1 PRF and HKDF. PBKDF1 and PKCS12 KDF are not implemented.

The low level KDF functions and PKEY KDF support have been changed to wrap the new API. Other parts of OpenSSL have not yet been updated to call the new API.

Documentation to follow.

- [x] documentation is added or updated
- [x] tests are added or updated

levitte commented on include/openssl/evp_kdf.h:
```diff
@@ -0,0 +1,117 @@
+# define EVP_KDF_CTRL_TLS_MD                    (EVP_KDF_ALG_CTRL + 4)
+# define EVP_KDF_CTRL_TLS_SECRET                (EVP_KDF_ALG_CTRL + 5)
+# define EVP_KDF_CTRL_TLS_SEED                  (EVP_KDF_ALG_CTRL + 6)
+# define EVP_KDF_CTRL_HKDF_MD                   (EVP_KDF_ALG_CTRL + 7)
```
Why not one generic `EVP_KDF_CTRL_MD`?

davidmakepeace replied:
That makes sense.
I made the change and added a generic `EVP_KDF_CTRL_ITER` as well.

paulidale commented:
We've got a single step KDF (SP 800-56) implemented and passing test vectors but it will need some cleanup to fit into this framework.  The intention is to submit this it after this PR has been accepted or at least after the final form of these changes are determined.

paulidale commented:
Thought from today: this interface should deal with NULL result buffers better.  Some of the algorithms use this to request the size of the result.  The rest are variably sized and don't support this functionality.  Uniformity would be better here.

My suggestion is that a new API is introduced that requests the size of the result (which matches digests) and that this returns 0 or -1 for variable size outcomes (I favour zero).  Then, the NULL checking can be removed from the new APIs and handed off to the legacy wrappers.

Thoughts?

t-j-h commented:
Do we really need the uint64 in the scrypt interface?

levitte commented:
The [scrypt page on tarsnap](http://www.tarsnap.com/scrypt.html) seems to indicate that at least `N` can be a large enough integer to warrant 64 bits...

t-j-h commented:
Sure it can technically go that high - but pragmatically would anyone ever exceed 2^32 - that I seriously doubt is realistic in *any* context

levitte commented:
@paulidale, it would be great if you could rebase and fix the conflicts.
(re crypto/err/openssl.txt, crypto/evp/evp_err.c, include/openssl/evperr.h and util/libcrypto.num, I would just toss them and do `make update` again)

paulidale commented:
I'll ask @davidmakepeace to update tomorrow.

levitte commented:
Oh oops, I mixed you guys up...  sorry :-/

paulidale commented:
I'm fatter and uglier.  I figured it would have been hard to mix us up :)

simo5 commented on crypto/evp/pkey_kdf.c:
```diff
@@ -0,0 +1,194 @@
+        cmd = EVP_KDF_CTRL_TLS_SEED;
+        break;
+    case EVP_PKEY_CTRL_HKDF_KEY:
+        cmd = EVP_KDF_CTRL_HKDF_KEY;
```
Key may be used in other KDFs (like SSH KDF), would it make sense to use a generic EVP_KDF_CTRL_KEY ?

levitte replied:
@davidmakepeace will have to decide...  my two öre (hey, Sweden) is that yes, this should happen, and more generally speaking, as soon as a control is shared among more than one specific KDF, the control should be generalised, *as long as* the passed value is the same (i.e. same fixed size, or variable size allowed)

simo5 replied:
@levitte I just commented on this because a "Key Derivation Function" usually has a Key (aka Secret) as input (and other data) and a derived key as output, that's all.

davidmakepeace replied:
Having a generic EVP_KDF_CTRL_KEY does make sense. I will make that change.

simo5 commented:
@davidmakepeace any chance you will rebase this PR soonish, it's been indicated as a base to be used to implement SSHKDF and it'd be nice to start on a fresh rebase that does not conflict with master

paulidale commented on include/openssl/evp_kdf.h:
```diff
@@ -0,0 +1,109 @@
+/*
```
To clarify a little.

Should the new APIs be moved to `kdf.h`?  My thought is yes.

I agree that the old #defines should be phrased in terms of the new ones, if possible.

Should the old APIs be moved into a different file or wrapped in a `#if OPENSSL_API_COMPAT` check?  I'm not sure.  I'm not keen on a second file unless we mark it as _going to go away eventually_ somehow.  Could we have a _deprecated_ sub-directory and `kdf.h` `#include <openssl/deprecated/old_kdf.h>`?

Or should the old APIs stay in `kdf.h` along with the new?

davidmakepeace replied:
I will move the new APIs into `kdf.h` and make the constants for the old API be `#defined` in terms of the ones for the new APIs. I will clearly delineate the boundary between the new and old APIs. We can deprecate and move the old API out to another header file later, once there is consensus on doing so.

davidmakepeace replied:
I moved the new APIs into `kdf.h`. Breaking binary compatibility would make it harder to back-port this to older versions, so the old cmd constants are not defined in terms of the new ones.

davidmakepeace commented on crypto/evp/evp_kdf.c:
```diff
@@ -0,0 +1,177 @@
+/*
```
Okay, I can do that.

Hmmm. Yeah, I am having trouble seeing much consistency. EVP_CIPHER_CTX_*() functions are defined in crypto/evp/evp_enc.c while EVP_MD_CTX_*() functions are defined in crypto/evp/digest.c . :)

simo5 commented on include/openssl/evp_kdf.h:
```diff
@@ -70,6 +70,9 @@ int EVP_KDF_derive(EVP_KDF_CTX *ctx, unsigned char *key, size_t *keylen);
 # define EVP_KDF_CTX_set_md(kctx, md) \
             EVP_KDF_CTX_ctrl(kctx, EVP_KDF_CTRL_MD, 0, (void *)(md))
 
+# define EVP_KDF_CTX_set1_key(kctx, key, keylen) \
```
Sorry for the possibly stupid question but why set1 instead of just set ?
I may be missing the context of what the 1 means here...

mattcaswell replied:
This is a common OpenSSL idiom. We should really write this down somewhere in the documentation!

get0 == get a value such that the pointer points to the internal data value. The caller should not free it
get1 == get a value such that the pointer is to memory that the caller should free
set0 == set a value and pass the ownership of the pointer to OpenSSL. Once successfully set the caller should not directly free it
set1 == set a value but retain ownership of the pointer with the caller, i.e. the caller should free it

kaduk replied:
I like to think of it as "0 (or 1) references/copies are created"

levitte commented:
This should be labeled 1.2.0 at this point, shouldn't it?

simo5 commented on doc/man3/EVP_KDF_CTX_set_hkdf_mode.pod:
```diff
@@ -0,0 +1,145 @@
+ size_t outlen = sizeof(out);
+ kctx = EVP_KDF_CTX_new_id(EVP_KDF_HKDF);
+
+ if (EVP_KDF_CTX_set_hkdf_md(kctx, EVP_sha256()) <= 0)
```
EVP_KDF_CTX_set_hkdf_md -> EVP_KDF_CTX_set_md

simo5 commented on doc/man3/EVP_KDF_CTX_set_hkdf_mode.pod:
```diff
@@ -0,0 +1,145 @@
+     /* Error */
+ if (EVP_KDF_CTX_set1_hkdf_salt(kctx, "salt", 4) <= 0)
+     /* Error */
+ if (EVP_KDF_CTX_set1_hkdf_key(kctx, "secret", 6) <= 0)
```
EVP_KDF_CTX_set1_hkdf_key -> EVP_KDF_CTX_set1_key

davidmakepeace replied:
Thanks!
You missed:
EVP_KDF_CTX_set1_hkdf_salt -> EVP_KDF_CTX_set1_salt
:)

simo5 commented on include/openssl/kdf.h:
```diff
@@ -10,29 +10,135 @@
+                             EVP_KDF_CTRL_SCRYPT_MAXMEM_BYTES, maxmem_bytes)
+
+
+/**** The old PKEY-based KDF API follows. ****/
```
So these defines make the API source level compatible, but breaks the ABI, is this OK ?
Why not retain the original values so that ABI is not broken ?

paulidale replied:
Some of the earlier constants have been merged in the new.  E.g. the salts.

A decision that the next OpenSSL release will be ABI breaking has been made, so this is okay.

levitte replied:
That shouldn't be too hard to fix.  Start with giving the new constants a different numeric range, then keep the old constants in an internal header file, renamed with `OLD_` in their names, and handle both the new and the old in the code where it makes sense.

levitte replied:
@paulidale, please note that formally speaking, we haven't actually made that decision yet.

simo5 replied:
I am actually planning to backport this to 1.1, so it would be nice if there weren't gratuitous breakages if at all possible. I know how to handle this myself if you need to stick with the current code though, just mentioning why I care.

paulidale replied:
If we rename the constants we're breaking the API -- the new names are different to the old.  The raised concern was that the constants were defined twice, rather than in terms of each other.

Back to defining them twice and staying ABI compatible, or define one in terms of the other ?

davidmakepeace replied:
I will put it back the way it was, preserving binary compatibility.

davidmakepeace replied:
I put the switch statement back that maps between the old and new cntrl codes and I moved the new codes to a different range so that it will force an error if the old constants are used with the new API.

simo5 commented:
@davidmakepeace anything still missing I can help you with ?

FdaSilvaYY commented on crypto/kdf/hkdf.c:
```diff
@@ -58,40 +58,41 @@ static int pkey_hkdf_init(EVP_PKEY_CTX *ctx)
         if (p1 < 0)
             return 0;
 
         if (kctx->salt != NULL)
```
useless check against NULL

FdaSilvaYY commented on crypto/kdf/scrypt.c:
```diff
@@ -9,15 +9,19 @@
 
 static int atou64(const char *nptr, uint64_t *result);
+static int scrypt_alg(const char *pass, size_t passlen,
+                  const unsigned char *salt, size_t saltlen,
```
nit: one indent missing  to align argument

simo5 commented on crypto/evp/kdf_lib.c:
```diff
@@ -0,0 +1,176 @@
+    return *ret;
+}
+
+EVP_KDF_CTX *EVP_KDF_CTX_new_id(int id)
```
Other EVP CTX functions are called just EVP_XYZ_CTX_new() why this one is _new_id() instead ?
I found it odd as I was trying to call EVP_KDF_CTX_new() and found it didn't exist...

davidmakepeace replied:
It takes a NID as a parameter and is modeled after the `EVP_PKEY_CTX_new_id(int id, ENGINE *e)` function, which is used for the existing PKEY KDF support, except that `EVP_KDF_CTX_new_id(int id)` doesn't take an `ENGINE*` parameter. Engine support is likely to be replaced with something else at some point in the future.

Many of the other `*_CTX_new()` functions take no parameters and the resulting context is not associated with any particular algorithm (i.e. NID).

Another question that could legitimately be asked is why the Derive function is called `EVP_KDF_derive()` instead of `EVP_KDF_CTX_derive()` or even `EVP_Derive()` (like `EVP_EncryptFinal()`). Well, the answer is that it is modeled after the `EVP_PKEY_derive()` function, which is used for the existing PKEY KDF support. I am not particularly attached to the current names and am happy to change it if there was some consensus on an alternative naming convention, though.

simo5 replied:
I forgot that EVP_PKEY_CTX_new_id() also used the _id() suffix ... I guess I still prefer just _new(), but it is simply a matter of personal taste.
EVP_KDF_derive() makes sense to me.

davidmakepeace replied:
There are several ways of identifying an algorithm - short name, long name, ASN1_OBJECT, NID - and I imagined that there could be other `EVP_KDF_CTX_new_*()` variants in the future.
I was not sure if it should be `*_new_id()`, `*_new_nid()` or `*_new_bynid()`, etc., but there was an existing `*_new_id()` so I went with that.

paulidale commented:
Ping?

richsalz commented:
Still a WIP?

levitte commented:
This is labeled 1.2.0.  Is that still current or is this API/ABI compatible with 1.1.1?

simo5 commented:
@levitte it's compatible with 1.1.1 afaict (I have a backport running on my machine with no ABI issues so far)

levitte commented:
Ok, I've relabeled it 'master' to reflect that

levitte commented on doc/man7/scrypt.pod:
```diff
@@ -86,26 +83,28 @@ This example derives a 64-byte long test vector using scrypt using the password
 =head1 COPYRIGHT
 
-Copyright 2017-2018 The OpenSSL Project Authors. All Rights Reserved.
+Copyright 2018 The OpenSSL Project Authors. All Rights Reserved.
```
Uhmmm, that isn't right

davidmakepeace replied:
Oops. Yeah, that started being written as a new file and ended up being almost the same as and replacing the existing file - except now the copyright notice is wrong.
I'll fix that.

paulidale commented:
To report errors at the EVP level, the lower level internal implementation would have to return which error occurred.  Currently, this is done rather inconsistently.  Could we standardise on something more meaningful than small integers or varying sign?  An enum return perhaps?

t-j-h commented:
No to any use of enums in the public API - they are a variable size concept - and hence non-ABI friendly.

paulidale commented:
That is precisely why I wrote _internal_.

t-j-h commented:
But errors need to propagate up through the API - so eventually the user gets some error indication. And using one system internally and another externally would seem to be a strange choice. 
I think you might be thinking about perhaps a different error reporting mechanism - as the caller is the one that needs to know what is going on - so ultimately the detail has to make it somewhere useful.

paulidale commented:
@levitte suggested pushing errors to the stack from the EVP layer and I was hoping that low level functions could return a more meaningful code than the `-2` mentioned.  enum { KDF_BAD_CTRL = -2 } would provide this.  A #define would too but for internal returns, an enum could provide additional checking -- compilers can provide coverage checking in switch statements.

Public APIs would have to return an `int` still, this would require at most a cast.

levitte commented:
I hadn't thought of all ramifications, frankly.  I just found it jarring that different EVP_KDF level control functions would treat the return from the implementation control function differently, and especially the handling of -2.

I was looking for a quick fix to get the treatment more similar at the EVP_KDF level here, not a full blown architectural change.  But between having every implementation have to set the same error in the stack, or just return -2 and let the EVP_KDF level take care of adding the error to the error stack, I prefer the latter.

davidmakepeace commented:
> 
> 
> Thought from today: this interface should deal with NULL result buffers better. Some of the algorithms use this to request the size of the result. The rest are variably sized and don't support this functionality. Uniformity would be better here.
> 
> My suggestion is that a new API is introduced that requests the size of the result (which matches digests) and that this returns 0 or -1 for variable size outcomes (I favour zero). Then, the NULL checking can be removed from the new APIs and handed off to the legacy wrappers.
> 
> Thoughts?

I like that idea. It would then be possible to change the third parameter of `EVP_KDF_derive()` to be an *in* parameter (`size_t`) instead of an *in/out* parameter (`size_t *`).

davidmakepeace commented on crypto/evp/kdf_lib.c:
```diff
@@ -0,0 +1,176 @@
+        return -2;
+    }
+    return ctx->kmeth->ctrl_str(ctx, name, value);
+}
```
I will change those functions to be more consistent. They are based on the corresponding PKEY functions in `crypto/evp/pmeth_lib.c`, which are also inconsistent.

levitte replied:
Yup, gotcha...  and yeah, unfortunately, there are inconsistencies in our current code.  Let's try not to reproduce them too much :-)

davidmakepeace replied:
I made those changes, including changing `EVP_KDF_CTX_ctrl_str()` to push the error.
Since kdf_lib.c is in `crypto/evp`, not `crypto/kdf`, it is now an EVP error instead of a KDF one.
I.e. there is a new reason code `EVP_R_UNKNOWN_PARAMETER_TYPE` and the old one, `KDF_R_UNKNOWN_PARAMETER_TYPE`, is no longer used.

Is that what you had in mind?

davidmakepeace commented on crypto/evp/kdf_lib.c:
```diff
@@ -0,0 +1,176 @@
+    if (len > INT_MAX)
+        return -1;
+    return ctx->kmeth->ctrl(ctx, cmd, len, (void *)str);
+}
```
Yes. I will fix that.

davidmakepeace replied:
I have removed NULL checks of pointers that should never be NULL.
All `EVP_KDF_CTX` objects have a method table, so I don't check if kmeth is NULL.
All KDFs require `ctrl` and `derive` methods, otherwise there is no way to provide any inputs or get output.
The `ctrl_str` and `output_size` methods are optional, so they are NULL-checked.

davidmakepeace commented on crypto/evp/kdf_lib.c:
```diff
@@ -0,0 +1,176 @@
+        rv = ctx->kmeth->ctrl(ctx, cmd, binlen, bin);
+    OPENSSL_free(bin);
+    return rv;
+}
```
I will fix that.

levitte commented:
Ok, so here's a question...  or rather an observation and food for thought.

Have you noticed that in *every* KDF method, you end up having to refer to `ctx->data` all the time, i.e. all methods get access to the whole EVP layer structure, but are really only interested in their own local structure?  This is something in the existing EVP constructs that I've observed before and found jarring.  We can't do much about what already exists until the next major version, but would anyone mind if we did that a bit differently for new stuff?  This is, after all, a completely new API, and we can form the method hooks exactly the way we want, so there's no formal reason to do it, well, better.

So on the EVP layer, instead of passing down `ctx` to the methods, `ctx->data` should be passed, and the method's `init` function should return `void*` rather than `int`.

Does this make sense?

simo5 commented:
OTOH having a generic context allows to set additional data hat is not specific to a KDF, and have it available in every function if needed later.
For example one may later devise an auditing/logging etc set of functions that need some handle in the ctx to do their thing.
If you pass down ctx->data, then you'll neet to add this handle in each specific function context or come up with ugly back pointers to ctx from within ctx->data

levitte commented:
Wouldn't such auditing/logging be better suited to be placed in the general EVP layer?

richsalz commented:
Doesn't C guarantee that the address of a struct is the same as the address of its first member?  To put the context first.

levitte commented:
Uhmmm, @richsalz, was that related to what I wrote?

richsalz commented:
Our edits crossesd in the aether. :)  I was commenting on everything getting a "ctx->data" pointer.  Just past the ctx pointer.

levitte commented:
You have misunderstood.  Passing `ctx` down to the methods is exactly what's happening now, which forces all the method functions to use `ctx->data` to get to the method data...  and that's basically all that the methods are (or should) be interested in.

richsalz commented:
If ctx->data == ctx, then you could just cast.

levitte commented:
But it isn't.  `ctx->data` is method local data, not EVP layer data.  Making them the same would be catastrophic.

richsalz commented:
Sorry to distract everyone; I was confused.

levitte commented on crypto/evp/kdf_lib.c:
```diff
@@ -0,0 +1,183 @@
+    ret = ctx->kmeth->ctrl_str(ctx, type, value);
+
+    if (ret == -2)
+        EVPerr(EVP_F_EVP_KDF_CTX_CTRL_STR, EVP_R_UNKNOWN_PARAMETER_TYPE);
```
I don't quite agree with this message.  -2 means that the control command isn't supported, and that's the error `EVP_KDF_CTX_ctrl` sets for the same return...  so I don't quite understand why this spot is different in that regard?  So unless you can argue for this message, I'd rather think it should be `ERR_R_COMMAND_NOT_SUPPORTED`.

davidmakepeace replied:
Yeah, I am happy to change that.

It was inherited from the current (PKEY) code which uses the `KDF_R_UNKNOWN_PARAMETER_TYPE` reason to indicate that the `ctrl_str()` type parameter (a string) is not recognized for the algorithm, as opposed to using a command that is not appropriate for the algorithm. It is subtly different from `EVP_R_COMMAND_NOT_SUPPORTED` since multiple types can map onto one command (e.g. "salt" and "hexsalt" both map on to the `EVP_KDF_CTRL_SALT` command). We probably don't need that distinction in the reason codes, though.

Note that I don't push any additional errors in the PKEY wrapper layer to ensure that the errors pushed onto the error stack exactly match the errors that are pushed by the current PKEY KDF implementation. Is that a problem? The return values will be the same, though.
What is the policy regarding changing the errors (reason codes) reported by existing APIs?
I had already changed from `KDF_R_UNKNOWN_PARAMETER_TYPE` to `EVP_R_UNKNOWN_PARAMETER_TYPE` when I moved the place where the error was pushed.

levitte commented:
> I wasn't sure if we would need access to the EVP_KDF_CTX from the algorithm implementations in the future, though we probably won't.

I would argue that the implementation can put whatever in its own private structure, no need to put things in the global structure.  There are other classes where we do that (for example, there's the running IV in EVP_CIPHER_CTX), but in all cases I've seen, I could argue that it's just a convenience and that those data fields could just as well have ended up with the implementation local data.

levitte commented:
> While I'm at it.. I had included the `EVP_KDF_CTX_str2ctrl()`, `EVP_KDF_CTX_hex2ctrl()`...

The contruction of those functions is unfortunately geared for use with the EVP layer structure, which makes them quite nasty to use inside an implementation if a stricter separation of data structures is desirable (and I'll keep arguing it is).  [lament: oh how I wish C had lambda functions with context!]

So for now, I'd say that they can be ditched, along with the `_ctrl_str` function...  if we find that we want to make KDFs configurable through a config module, we can always recreate these functions later.

(meanwhile, I'm working on a generic replacement for `EVP_PKEY_CTX_str2ctrl` and `EVP_PKEY_CTX_hex2ctrl` that takes a function to be called with the resulting value.  That should be enough to create a lambda like functionality.

levitte commented on crypto/kdf/hkdf.c:
```diff
@@ -284,12 +253,10 @@ static unsigned char *HKDF_Expand(const EVP_MD *evp_md,
     unsigned char *ret = NULL;
 
```
```suggestion
-
```

levitte commented on crypto/kdf/kdf_local.h:
```diff
@@ -0,0 +1,17 @@
+int KDF_hex2ctrl(void *ctx, int (*ctrl)(void *ctx, int type, int p1, void *p2),
+                 int cmd, const char *hex);
+int KDF_md2ctrl(void *ctx, int (*ctrl)(void *ctx, int type, int p1, void *p2),
+                int cmd, const char *md);
```
We traditionally lowercase internal symbols...

davidmakepeace replied:
Done.

levitte commented:
> I changed the KDF implementations to get a pointer to their own structure instead of an EVP_KDF_CTX pointer.

Nicely done

davidmakepeace commented on crypto/kdf/hkdf.c:
```diff
@@ -284,12 +253,10 @@ static unsigned char *HKDF_Expand(const EVP_MD *evp_md,
     unsigned char *ret = NULL;
 
```
Done.

davidmakepeace commented:
> @paulidale said:
> 
> > Thought from today: this interface should deal with NULL result buffers better. Some of the algorithms use this to request the size of the result. The rest are variably sized and don't support this functionality. Uniformity would be better here.
> > My suggestion is that a new API is introduced that requests the size of the result (which matches digests) and that this returns 0 or -1 for variable size outcomes (I favour zero). Then, the NULL checking can be removed from the new APIs and handed off to the legacy wrappers.
> > Thoughts?
> 
> I like that idea. It would then be possible to change the third parameter of `EVP_KDF_derive()` to be an _in_ parameter (`size_t`) instead of an _in/out_ parameter (`size_t *`).

I went ahead and added a `EVP_KDF_CTX_output_size()` function. `EVP_KDF_CTX_derive()` can no longer be used to query the output size by passing a NULL output buffer.
I left the the third parameter as an _in/out_ parameter (`size_t *`).
(I also renamed `EVP_KDF_derive()` to `EVP_KDF_CTX_derive()` since it looks strange to have a `EVP_KDF_CTX` method be missing the `_CTX`. Yes, I know some existing APIs do that.)

Naturally the `EVP_PKEY_derive()` method still allows the output size to be queried by passing NULL for the buffer. It is emulated by calling `EVP_KDF_CTX_output_size()`.

levitte commented on doc/man3/EVP_KDF_CTX_new_id.pod:
```diff
@@ -0,0 +1,111 @@
+
+ size_t EVP_KDF_CTX_output_size(EVP_KDF_CTX *ctx);
+
+ int EVP_KDF_CTX_derive(EVP_KDF_CTX *ctx, unsigned char *key, size_t *keylen);
```
Another few nits:

- it would be nice if `EVP_KDF_CTX` was mentioned here, with the typedef from `ossl-typ.h` in the SYNOPSIS (we usually place it right after the `#include` line) and the name itself in the NAME section.  That way, people who want to start somewhere will be able to just `man EVP_KDF_CTX`.

- We usually don't have empty lines between every function in the SYNOPSIS, except possibly for grouping when there's a huge number of functions described (and even then, not always)...

davidmakepeace replied:
No worries - I am still learning what the conventions are.
Thanks for the guidance!

simo5 commented on crypto/evp/kdf_lib.c:
```diff
@@ -132,12 +132,19 @@ int EVP_KDF_CTX_ctrl_str(EVP_KDF_CTX *ctx,
 
+size_t EVP_KDF_CTX_output_size(EVP_KDF_CTX *ctx)
+{
+    if (ctx == NULL || ctx->kmeth->output_size == NULL)
```
Wouldn't it be better to return an error if ctx is NULL ?
Given 0 is used to signifiy variable length, the return value would be wrong as a generic return value ...

davidmakepeace replied:
Yes, 0 can mean either variable length or an error.

What value would you suggest returning in the case of an error, then? `SIZE_MAX`?
We could also change the return type to either `int` or `ossl_ssize_t` and return -1.

Alternatively, we could just say that attempting to get the output size for a variable-output-size algorithm is an error and therefore 0 always represents an error.

Thoughts?

Currently only HKDF in EVP_KDF_HKDF_MODE_EXTRACT_ONLY will produce a fixed output size.
An error can occur if either `ctx` is `NULL` or the `MD` has not yet been set.

simo5 replied:
I do not know what's the best way to proceed, I guess "just say that attempting to get the output size for a variable-output-size algorithm is an error" may be a good way to put it ...

davidmakepeace replied:
We could also remove the `NULL` check for `ctx` and state that returning 0 for `MD` not being set just indicates that the output size is not yet known. I.e. errors don't actually occur.

Another alternative... return SIZE_MAX for variable output size and 0 for errors.

davidmakepeace replied:
After giving it some thought I ended up using SIZE_MAX to indicate variable output size and 0 for errors.

stevelinsell commented on crypto/kdf/tls1_prf.c:
```diff
@@ -31,145 +32,127 @@ typedef struct {
     unsigned char seed[TLS1_PRF_MAXBUF];
     size_t seedlen;
-} TLS1_PRF_PKEY_CTX;
+} TLS1_PRF_KDF_CTX;
```
I'm a little worried about this change. The QuickAssist Engine that I maintain provides an implementation of PRF for offloading to hardware. As part of this we need to maintain our own ctx info related to the request.
Currently with the old EVP_PKEY_CTX when we provide our own init, cleanup, ctrl and derive functions we can specify our own internal data structure and assign it to the ctx->data structure that was a void pointer and include our own data structures. As we provide our own init and cleanup functions we had full control as long as we provided our own functions for everything that uses the data structure. With this change we now have a tightly defined TLS1_PRF_KDF_CTX data structure with no field for user defined data. Can we at least have some kind of pointer (perhaps names custom) in the structure that is a void pointer that engines can use to attach their own data to. The same also applies to the HKDF_KDF_CTX and any other KDF CTX structures.

stevelinsell replied:
Actually am I misunderstand the implementation, as a void pointer is always passed around, does that mean I can define my own data structure within my engine implementation and always cast the ctx to my data structure? If so I take back what I said, the implementation is fine.

levitte replied:
I'm not sure I understand everything you said there, but it basically sounds like you had your own EVP_PKEY_METHOD.  With this new API, you will have to provide a EVP_KDF_METHOD instead.

davidmakepeace replied:
This is a work-in-progress. Currently there is no way for an engine or an application to provide an EVP_KDF_METHOD. At some point there will be a way to do that; whether it be via engine or some other mechanism is not yet clear.

richsalz commented:
Should the functions have get/set explicitly in the name? For example SHA3 is variable-sized.

davidmakepeace commented:
> 
> 
> Should the functions have get/set explicitly in the name? For example SHA3 is variable-sized.

I guess you mean so the output size can be set.

BTW, SHA-3 is only variable-sized if you use the SHAKE variants (XOF).

I don't think we will need to set the output size in the case of a KDF using SHA3 (SHAKE) underneath because that will be a variable-output-size KDF and the amount of output will be determined by the size of the output buffer given to the derive method.

levitte commented on crypto/evp/kdf_lib.c:
```diff
@@ -0,0 +1,150 @@
+    return ctx->kmeth->output_size(ctx->data);
+}
+
+int EVP_KDF_CTX_derive(EVP_KDF_CTX *ctx, unsigned char *key, size_t *keylen)
```
So @paulidale is talking to me about avoiding the keylen pointer thing and to only have users find out the output size through the dedicated function for it (`EVP_KDF_CTX_output_size` in your case).

Since you and I are current working on fairly similar things and battling the same coding patterns, I think it would be nice if we had some similarity in the resulting API.

I'm up for a discussion for this case.  Pass a pointer to a `size_t` and have it filled, or pass just a `size_t` and have the derivation function return with an error if it's wrong (and is larger than necessary wrong?)

simo5 replied:
Given keylen does not need to convey both buffer size and return (once you have a way to query for output_size if the length can't be arbitrary) then I agree that no passing a pointer is a better API, less chances for mistakes.

davidmakepeace replied:
I got rid of the keylen pointer thing. It is now a size_t.
The way I have implemented it, for fixed-output algorithms (or modes) keylen must exactly match the output size or an error results.

richsalz commented:
We have get/set all over the place.  I think you should reconsider.

davidmakepeace commented:
@richsalz  said:
> 
> We have get/set all over the place. I think you should reconsider.

Do you have any specific examples.

These ones don't have *get* in their name:
`EVP_MD_size()`, `EVP_MD_block_size()`, `EVP_MD_CTX_size()`, `EVP_MD_CTX_block_size()`,
`EVP_CIPHER_block_size()`, `EVP_CIPHER_key_length()`, `EVP_CIPHER_iv_length()`,
`EVP_CIPHER_CTX_block_size()`, `EVP_CIPHER_CTX_key_length()`, `EVP_CIPHER_CTX_iv_length()`,
`EVP_PKEY_size()`, and now: `EVP_MAC_size()`.

davidmakepeace commented:
I plan to update this PR to bring it into alignment with conventions used in the new EVP_MAC API.

levitte commented:
> I plan to update this PR to bring it into alignment with conventions used in the new EVP_MAC API.

Looking forward to it :-)

richsalz commented:
```
; grep _[gs]et_ util/lib*.num  | wc   
   1227    4908  107348
```

levitte commented:
Yes, we have explicit getters and setters, mostly when they come in pairs. It makes no real sense to have an explicit `_get_` in the name when there's no corresponding setter and most likely will not be.

daurnimator commented on doc/man3/EVP_KDF_CTX_set_hkdf_mode.pod:
```diff
@@ -0,0 +1,149 @@
+     /* Error */
+ if (EVP_KDF_CTX_set1_salt(kctx, "salt", 4) <= 0)
+     /* Error */
+ if (EVP_KDF_CTX_set1_key(kctx, "secret", 6) <= 0)
```
Call this "key" instead of "secret"?

daurnimator replied:
I commented on the man page because it has all the names in a close area.
However the point I was trying to make was that we should be consistent, ideally all the five will match:

  - the function name (`"key"` in this example)
  - the `CTRL` define (`EVP_KDF_CTRL_KEY`)
  - the prefix used by `CTRL_STR` (`key` and `hexkey`)
  - the local variable name used in examples (`secret` in this example)
  - the name used in tests (seems to be `IKM` for HKDF tests)

When they don't match, it's easy to get confused between pass/password/key/secret, as all of these are present.

davidmakepeace replied:
`"secret"` is not a local variable name here, it is six bytes of sample data.

daurnimator replied:
To prevent confusion with `tls1_prf_secret` could you use `"mykey"` or something less ambiguous instead?

daurnimator commented on doc/man3/EVP_KDF_CTX_set_hkdf_mode.pod:
```diff
@@ -0,0 +1,149 @@
+     /* Error */
+ if (EVP_KDF_CTX_set1_key(kctx, "secret", 6) <= 0)
+     /* Error */
+ if (EVP_KDF_CTX_add1_hkdf_info(kctx, "label", 5) <= 0)
```
Should `EVP_KDF_CTX_add1_hkdf_info` be renamed to something containing "label"? e.g. `EVP_KDF_CTX_add1_label`

daurnimator replied:
That would be better! "label" is used in nearby locations with a different meaning. e.g. https://github.com/openssl/openssl/pull/6674/files#diff-4e6407b0ec75b41bf1eb29a8ef274bdbR28
```
Ctrl.label = seed:key expansion
```

davidmakepeace replied:
Well, `info` here is the name used in `HKDF` (RFC 5869), while `"label"` is just five bytes of sample data that makes reference to how it is used in TLS 1.3 (RFC 8446), a well-known user of `HKDF`.
Earlier TLS versions use `TLS1-PRF` and supply the label as the salt parameter to that KDF.

daurnimator commented on include/openssl/kdf.h:
```diff
@@ -10,10 +10,94 @@
+# define EVP_KDF_CTX_set_scrypt_p(kctx, p) \
+            EVP_KDF_CTX_ctrl_uint64(kctx, EVP_KDF_CTRL_SCRYPT_P, p)
+
+# define EVP_KDF_CTX_set_scrypt_maxmem_bytes(kctx, maxmem_bytes) \
```
Should `EVP_KDF_CTX_set_scrypt_maxmem_bytes` be more generic to cater for future algorithms that also need a memory cap?
Rename to `EVP_KDF_CTX_set_maxmem_bytes`?

davidmakepeace replied:
Yes, I can imagine other algorithms needing a memory cap.

davidmakepeace replied:
I made that parameter generic.

daurnimator commented on crypto/kdf/tls1_prf.c:
```diff
@@ -31,145 +32,127 @@ typedef struct {
         return 1;
 
-    case EVP_PKEY_CTRL_TLS_SEED:
+    case EVP_KDF_CTRL_TLS_SEED:
```
(excuse my lack of background knowledge here:) Is a TLS seed sufficiently different from a "salt"? (likewise is the secret sufficiently different from a "key"?): could they share the same define?

davidmakepeace replied:
I haven't checked to see if seed/salt can be considered equivalent here. Sometimes if the standard that defines an algorithm uses a particular terminology, it is good to be able to set a parameter using the same name as used in the standard - that way you know you are setting the parameter you think you are setting.
Also, sometimes algorithms that purport to solve the same problem are not interchangeable as they and their parameters have slightly different security properties; in that case it would be good for the parameters to have different names, even if at first glance they seem to be doing the same thing.

daurnimator replied:
From a binding perspective, it's good to have only a limited number of functions: if a new algorithm came out in a new openssl release that didn't require a different function call, then my users would instantly gain support.
However if every new algorithm requires some new `_set_egg` that is almost the same as `set_seed` then users have to wait for my bindings to update.

davidmakepeace replied:
New KDF algorithms do not add new function entry points to libcrypto.  `EVP_KDF_set1_salt()` and `EVP_KDF_add1_tls1_prf_seed()` are just macros provided for convenience that invoke the same function, `EVP_KDF_ctrl()`; you don't have to use them.
You could call either `EVP_KDF_ctrl()` or `EVP_KDF_ctrl_str()` directly if you want to avoid new bindings when new algorithms are added. In particular, `EVP_KDF_ctrl_str()` is designed to allow applications to supply parameters as strings (obtained from a config file or the command line, for example) so that new algorithms can be supported without even recompiling the application.

Note that in this case `EVP_KDF_set1_salt()` and `EVP_KDF_add1_tls1_prf_seed()` are different in at least one major way: `EVP_KDF_add1_tls1_prf_seed()` appends the bytes to the algorithm's internal buffer while `EVP_KDF_set1_salt()` completely replaces the existing value.

davidmakepeace replied:
All the parameter-setting macros are gone now, so from a binding perspective there are a limited number of functions and new ones won't be added as new algorithms are added.

daurnimator commented on doc/man3/EVP_KDF_CTX_set_hkdf_mode.pod:
```diff
@@ -0,0 +1,149 @@
+
+=over 4
+
+=item EVP_KDF_HKDEF_MODE_EXTRACT_AND_EXPAND
```
This says `EVP_KDF_HKDEF_MODE_EXTRACT_AND_EXPAND` rather than `EVP_KDF_HKDF_MODE_EXTRACT_AND_EXPAND` (note extra 'E')

davidmakepeace replied:
Thanks!

davidmakepeace commented on doc/man3/EVP_KDF_CTX_set1_pass.pod:
```diff
@@ -0,0 +1,85 @@
+ int EVP_KDF_CTX_set1_salt(EVP_KDF_CTX *kctx, unsigned char *salt,
+                           int saltlen);
+
+ int EVP_KDF_CTX_set_iter(EVP_KDF_CTX *kctx, const EVP_MD *md);
```
Thanks!

daurnimator commented:
How likely to be merged is this currently proposed API?
I've written https://github.com/wahern/luaossl/pull/154 which backports the API in this PR to older OpenSSL versions.

simo5 commented:
@daurnimator the API seem to be still seeing some minor changes as reviewers go through it.
I have a similar backport ready plus changes in other libraries waiting on this API to be committed (and hopefully not changed after this PR goes in).

levitte commented:
Dudes, I would advise a little bit of patience.  A previous message was that @davidmakepeace was going to align the EVP_KDF API along the EVP_MAC ideas, and that does require a bit of rethinking.  Whether that leads to very large changes compared to what's in this PR now or not, I cannot say.  I suspect not, I assume the changes will be subtle, but important.

levitte commented on crypto/include/internal/evp_int.h:
```diff
@@ -143,6 +143,30 @@ extern const EVP_MAC poly1305_meth;
+    const EVP_KDF_METHOD *kmeth;
+    /* Algorithm specific data */
+    EVP_KDF_IMPL *impl;
+} /* EVP_KDF_CTX */ ;
```
I would move this context to `crypto/evp/evp_locl.h`

levitte commented on crypto/kdf/kdf_util.c:
```diff
@@ -0,0 +1,73 @@
+        return 0;
+    }
+    return call_ctrl(ctrl, impl, cmd, md);
+}
```
I wouldn't mind at all if these got generalised into the public API.  Suggested names: `EVP_vstr2ctrl`, `EVP_vhex2ctrl`, `EVP_vmd2ctrl`.

davidmakepeace replied:
Hmmm. I don't think we need them yet, at least not for KDF. They wouldn't be needed to help with implementing KDF algorithms as there is currently no way to do that.

I have tried to avoid adding any brand new legacy stuff, such as ENGINE support, since we don't know how the proposed provider model will pan out. There is currently no way to set methods on an `EVP_KDF_METHOD`, for example, and `EVP_KDF_METHOD` isn't even defined in `include/openssl/evp.h` (or any public header). That's mainly because if/when it does become public it will probably be called `EVP_KDF`, analogous to the new `EVP_MAC`.

levitte replied:
I'm not sure how ENGINE came into play...  I was rather thinking of function that could join the already implemented `EVP_str2ctrl` and `EVP_hex2ctrl`, the only difference being that these are more elegantly suited for use with variadic style control functions.

davidmakepeace replied:
Yeah, okay. I was just trying to guess the use case for those being in the public API.

davidmakepeace replied:
@levitte , why did you add `EVP_str2ctrl` and `EVP_hex2ctrl` to the public API?

levitte replied:
Because I saw a general usefulness...  I don't think we're getting rid of control functions any time soon, regardless of anything else, so I believe these kinds of utility functions have their place there.

levitte commented on doc/man3/EVP_KDF_set1_pass.pod:
```diff
@@ -0,0 +1,78 @@
+ int EVP_KDF_set_iter(EVP_KDF_CTX *kctx, int iter);
+ int EVP_KDF_set_md(EVP_KDF_CTX *kctx, const EVP_MD *md);
+ int EVP_KDF_set1_key(EVP_KDF_CTX *kctx, const unsigned char *key,
+                      size_t keylen);
```
You call the context pointer `ctx` in `EVP_KDF_CTX.pod`, but `kctx` here...  any reason why?

davidmakepeace replied:
There is no good reason. The ones with `kctx` (in several pod files) are documenting macros while `EVP_KDF_CTX.pod` is documenting real functions. I will make them more consistent.

davidmakepeace replied:
Done.

levitte commented:
Oooh, having direct KDF tests and PKEY KDF tests entirely separate was a good idea!  Why didn't I think of that for EVP_MAC?

levitte commented on crypto/err/openssl.txt:
```diff
@@ -740,6 +740,9 @@ EVP_F_EVP_DIGESTFINALXOF:174:EVP_DigestFinalXOF
 EVP_F_EVP_DIGESTINIT_EX:128:EVP_DigestInit_ex
 EVP_F_EVP_ENCRYPTFINAL_EX:127:EVP_EncryptFinal_ex
 EVP_F_EVP_ENCRYPTUPDATE:167:EVP_EncryptUpdate
+EVP_F_EVP_KDF_CTRL:217:EVP_KDF_ctrl
```
`make update` tells us that this number conflicts with another `_F_` macro.

I usually resolve this by resetting error number carrying files to what they were pre-PR and run a new `make update`

davidmakepeace replied:
Yeah, I usually do that except when a `make update` doesn't modify any of those files relative to the previous commit.
I will try again after resetting them to the pre-PR state...

davidmakepeace replied:
Done.

davidmakepeace commented:
@levitte said:
> The [scrypt page on tarsnap](http://www.tarsnap.com/scrypt.html) seems to indicate that at least `N` can be a large enough integer to warrant 64 bits...

I will change the other parameters (r, p) from uint64_t to being uint32_t.

davidmakepeace commented:
Should the parameter-setting macros be replaced with inline functions?
That could be a question for another day, though. I guess it is fine for them to be macros for the moment.

Comments?

paulidale commented:
The Travis failure is not relevant.

levitte commented:
Regarding the `EVP_KDF` parameter-setting macros, you do have the freedom not to define them at all.  I think that's just as legitimate a choice as choosing the macro or inline form.

Just had to say that so we can remove the pressure of *having* to add them.

davidmakepeace commented:
> Regarding the `EVP_KDF` parameter-setting macros, you do have the freedom not to define them at all. I think that's just as legitimate a choice as choosing the macro or inline form.
> 
> Just had to say that so we can remove the pressure of _having_ to add them.

Yes, true, although `EVP_KDF_ctrl()` now being variadic requires a greater degree of discipline to not pass an `int` where a `size_t` or `uint64_t` is required, for example, especially now that platforms where `size_t` is larger than `int` are becoming more commonplace.

Despite `printf()` being variadic it at least gets some type checking of the parameters thanks to modern compilers. We don't have that luxury.
Using inline functions has the advantage over macros that there is some type checking rather than blindly casting anything to the required type, such as pointers to integers and vice versa.

levitte commented:
> Yes, true, although `EVP_KDF_ctrl()` now being variadic requires a greater degree of discipline to not pass an `int` where a `size_t` or `uint64_t` is required, for example, especially now that platforms where `size_t` is larger than `int` are becoming more commonplace.

The dangers there are often not as bad as one might thing, as most of these simple types usually fit in one stack slot, so they are in practice cast to something with whatever amount of bits the architecture allows.  On 64-bit architectures, the types you mention should therefore be less problematic than one might expect...

It's true that inline functions do provide the comfort of type safety (and most of all, correct order), so if you *must*, I would urge you to use those rather than macros...  (speaking of which, there's a macro `ossl_inline` which you should use for making inline macros)

davidmakepeace commented:
I will remove the macros and update the documentation.

paulidale commented:
Travis failures are not relevant.  The conflicts are.

davidmakepeace commented:
I will remove the macros and update the documentation.


> 
> 
> Travis failures are not relevant. The conflicts are.

I rebased just now.
I also removed the parameter-setting macros and have started updating the documentation.
More documentation updates to follow...

levitte commented:
> More documentation updates to follow...

I assume it's the documentation of the other KDFs that are missing in `doc/man7/`?

Other than that, I like what I'm looking at (preliminary approval)

davidmakepeace commented:
> > More documentation updates to follow...
> 
> I assume it's the documentation of the other KDFs that are missing in `doc/man7/`?

Yes, that's right. I'll get them done when I get some time to do that. I wanted to rebase and make the code changes available.

> Other than that, I like what I'm looking at (preliminary approval)

Great!

daurnimator commented:
What are the plans for macros and/or inline functions? Merge this first and add later?

davidmakepeace commented:
> What are the plans for macros and/or inline functions? Merge this first and add later?

The current plan is not to have either of them.

davidmakepeace commented:
@paulidale, @levitte I finally got around to hammering the documentation into shape.

Please take a look. Any additional functionality can be added after this is merged.

davidmakepeace commented:
ping

levitte commented:
I just had a quick look, and like what I see (and am noticing a documentation structure around controls that might inspire me to do something similar for EVP_MAC).  I want to take a second more thorough look before rubber stamping, and will ask for a bit of patience, as I'm being sucked up in meetings this week.

davidmakepeace commented:
> I just had a quick look, and like what I see (and am noticing a documentation structure around controls that might inspire me to do something similar for EVP_MAC). I want to take a second more thorough look before rubber stamping, and will ask for a bit of patience, as I'm being sucked up in meetings this week.

Yeah, no problem. I expected you would be tied up in a few meetings this week. :)

I just noticed another inconsistency between EVP_KDF and EVP_MAC.
`EVP_KDF_ctrl_str()` uses a type string of "md" to set a `EVP_MD *` using `EVP_KDF_CTRL_SET_MD` (for `HKDF` and `PBKDF2`), while `EVP_MAC_ctrl_str()` uses a type string of "digest" to set a `EVP_MD *` using `EVP_MAC_CTRL_SET_MD`.
EVP_PKEY_CTX_ctrl_str() uses both; "digest" is handled in `EVP_PKEY_CTX_ctrl_str()` while "md" is handled in the `HKDF` implementation. This PR doesn't change those semantics.

Hmmm. Maybe I should add "digest" for `HKDF` and `PBKDF2` under EVP_KDF and possibly remove "md" for `PBKDF2` (it has to stay for `HKDF`).  What do you think?

daurnimator commented on include/openssl/kdf.h:
```diff
@@ -10,10 +10,48 @@
+# define EVP_KDF_CTRL_SET_MD            0x04 /* EVP_MD * */
+# define EVP_KDF_CTRL_SET_KEY           0x05 /* unsigned char *, size_t */
+# define EVP_KDF_CTRL_SET_MAXMEM_BYTES  0x06 /* uint64_t */
+# define EVP_KDF_CTRL_SET_TLS_SECRET    0x07 /* unsigned char *, size_t */
```
Could this be renamed to remove the `TLS_` prefix? likewise for seed below.
When using the string ctrl form they are just `"secret"` and `"seed"`.

davidmakepeace commented on crypto/kdf/tls1_prf.c:
```diff
@@ -31,145 +35,134 @@ typedef struct {
+        len = va_arg(args, size_t);
+        OPENSSL_clear_free(impl->sec, impl->seclen);
+        OPENSSL_cleanse(impl->seed, impl->seedlen);
+        impl->seedlen = 0;
```
That behavior is inherited from `EVP_PKEY_CTX_set_tls1_prf_secret()`, is documented (see: doc/man3/EVP_PKEY_CTX_set_tls1_prf_md.pod for EVP_PKEY and doc/man7/EVP_KDF_TLS1_PRF.pod for EVP_KDF) and appears to be intentional, possibly so that the context seed buffer could be reset without having to call `EVP_PKEY_derive_init()` again.

Remember that `EVP_PKEY_CTX_add1_tls1_prf_seed()` (as well as `EVP_KDF_ctrl(kctx, EVP_KDF_CTRL_ADD_TLS_SEED, ...)`) appends to whatever is in the context seed buffer rather than replacing it so there has to be some way to reset it to an empty buffer for the case where the context is to be reused. I don't know the history behind it, but resetting the context seed buffer when setting the secret value is set is the solution that was chosen. (Therefore you have to set the secret value before the first chunk of the seed value.) Of course we now have a `EVP_KDF_reset()` function that will reset everything.

I know you want a clean abstraction in luaossl and probably have crypto-agility in mind. The EVP_KDF API supports several KDF and PRF algorithms that are not equivalent or interchangeable; some, such as the TLS PRF, are special purpose or only used for a particular protocol so there is no need for them all to work exactly the same as each other.

I will have a look at the code in libssl where this algorithm is used and think about what you have suggested.

@levitte what do you think? Is this something that we should change?

levitte replied:
When I designed `EVP_MAC`, I attempted to do so with a clear mind of what it should do and not let the `EVP_PKEY` mac implementations influence me too much.  In essence, the `EVP_PKEY` implementations had to adapt to the `EVP_MAC` API, not the other way around.

So I guess that's my best advice, to try and make `EVP_KDF` into something that makes sense in itself.  If you feel that it doesn't really make sense for it to reset the seed here, then don't!  Finding it hard to justify this behavior is an indicator that you might want to change that and then have the `EVP_PKEY` implementation do an extra dance, so to say.  `EVP_KDF_reset` might be useful, depending on how much it resets.

Throwing in another idea, you might want to consider extra controls, such as `EVP_KDF_CTRL_RESET_TLS_SEED`, which the `EVP_PKEY` implementation can then call appropriately.

levitte replied:
Now that I think of it, I would add a RESET control for any control that *adds* information, just for completeness and to give the users a sense of control.

davidmakepeace replied:
> Now that I think of it, I would add a RESET control for any control that _adds_ information, just for completeness and to give the users a sense of control.

Good idea. I will add RESET controls to go along with EVP_KDF_CTRL_ADD_TLS_SEED and EVP_KDF_CTRL_ADD_HKDF_INFO.

davidmakepeace replied:
I have made the change to add RESET controls; `EVP_KDF_ctrl(ctx, EVP_KDF_CTRL_SET_TLS_SECRET, ...)` no longer resets the context seed buffer so the commit fixes #7728 .

davidmakepeace commented:
Rebased to master.

simo5 commented:
Happy New Year!
What's the status with this PR?

FdaSilvaYY commented:
Just needs a Copyright update to 2019 ;)
I suggest you add a "[extended tests]" line in the last commit message to trigger a larger CI test coverage.

FdaSilvaYY commented on test/evp_kdf_test.c:
```diff
@@ -0,0 +1,213 @@
+
+    if (EVP_KDF_ctrl(kctx, EVP_KDF_CTRL_SET_MD, EVP_sha256()) <= 0) {
+        TEST_error("EVP_KDF_CTRL_SET_MD");
+        return 0;
```
should be a goto err; to allow `kctx` object release

davidmakepeace replied:
I have updated the tests as you suggested since they also serve to demonstrate how the API should be used.

FdaSilvaYY commented on test/evp_kdf_test.c:
```diff
@@ -0,0 +1,213 @@
+            return 0;
+        }
+    }
+    EVP_KDF_CTX_free(kctx);
```
Better rewrite as this  
```
    ret = 1
err:
    EVP_KDF_CTX_free(kctx);
    return ret;
```

FdaSilvaYY commented on test/evp_kdf_test.c:
```diff
@@ -0,0 +1,213 @@
+            return 0;
+        }
+    }
+    EVP_KDF_CTX_free(kctx);
```
idem

FdaSilvaYY commented on test/evp_kdf_test.c:
```diff
@@ -0,0 +1,213 @@
+            return 0;
+        }
+    }
+    EVP_KDF_CTX_free(kctx);
```
idem

FdaSilvaYY commented on test/evp_kdf_test.c:
```diff
@@ -0,0 +1,213 @@
+            return 0;
+        }
+    }
+    EVP_KDF_CTX_free(kctx);
```
idem

FdaSilvaYY commented on test/evp_test.c:
```diff
@@ -1871,13 +1871,14 @@ static const EVP_TEST_METHOD encode_test_method = {
     encode_test_run,
 };
 
+
```
useless change.

davidmakepeace replied:
The extra blank line was to match the extra spacing used elsewhere, such as before BASE64 TESTS.

FdaSilvaYY commented:
Ping @davidmakepeace : your PR needs few fixes and a quick rebase ;)

simo5 commented:
@davidmakepeace do you have plans to continue working on this PR ? Asking given the protracted silence.

davidmakepeace commented:
> Ping @davidmakepeace : your PR needs few fixes and a quick rebase ;)

Done.

davidmakepeace commented:
> 
> 
> @davidmakepeace do you have plans to continue working on this PR ? Asking given the protracted silence.

Yes.
Rebased.

levitte commented:
I just looked at the Travis failure, and that gets me a bit worried:

```
...
PYTHONPATH=../util VALGRIND="" python ./t_renprinc.py 
*** Failure: /home/travis/build/openssl/openssl/krb5/src/clients/kinit/kinit failed with code 1.
*** Last command (#8): /home/travis/build/openssl/openssl/krb5/src/clients/kinit/kinit v4
*** Output of last command:
Password for v4@KRBTEST.COM: 
kinit: Password incorrect while getting initial credentials

For details, see: /home/travis/build/openssl/openssl/krb5/src/tests/testlog
Or re-run this test script with the -v flag:
    cd /home/travis/build/openssl/openssl/krb5/src/tests
    PYTHONPATH=/home/travis/build/openssl/openssl/krb5/src/util /usr/bin/python ./t_renprinc.py -v

Use --debug=NUM to run a command under a debugger.  Use
--stop-after=NUM to stop after a daemon is started in order to
attach to it with a debugger.  Use --help to see other options.
make[3]: *** [check-pytests] Error 1
make[3]: Leaving directory `/home/travis/build/openssl/openssl/krb5/src/tests'
make[2]: *** [check-recurse] Error 1
make[2]: Leaving directory `/home/travis/build/openssl/openssl/krb5/src'
../../util/shlib_wrap.sh ../../test/recipes/95-test_external_krb5_data/krb5.sh => 2
not ok 1 - running krb5 tests

#   Failed test 'running krb5 tests'
#   at ../test/recipes/95-test_external_krb5.t line 23.
# Looks like you failed 1 test of 1.
Dubious, test returned 1 (wstat 256, 0x100)
Failed 1/1 subtests 
../test/recipes/95-test_external_pyca.t ....... 
1..1
```

I'll try and dive into this later this evening

levitte commented:
A look at `krb5/src/tests/testlog` on my machine suggests there's an issue with krbv4 tickets.  This is the relevant failure I find:

```
*** [7] Executing: .../krb5/src/kadmin/cli/kadmin.local addprinc -e aes128-cts:v4 -pw v417065 v4
*** [7] Completed with return code 0
*** [8] Executing: .../krb5/src/clients/kinit/kinit v4
Password for v4@KRBTEST.COM: 
kinit: Password incorrect while getting initial credentials
*** [8] Completed with return code 1
```

Should anyone be curious, [this is the test script that fails](https://github.com/krb5/krb5/blob/master/src/tests/t_renprinc.py)

simo5 commented:
@levitte I do not think we should care for Krb v4 at all anymore, might as well just remove anything v4 related from tests if you agree.

mattcaswell commented:
> @levitte I do not think we should care for Krb v4 at all anymore, might as well just remove anything v4 related from tests if you agree.

Well these are external tests imported from the krb5 project - so unless we were to remove the whole test suite (which we don't want to do), it's not for us to say (nor in our power) to say whether they stay or go. More importantly a test failure indicates a possible regression in the way OpenSSL behaves that might impact other applications - so we're interested in understanding what that regression is.

simo5 commented:
@mattcaswell that makes sense, I'll put my krb hat on and go investigate upstream about these v4 tests.

greghudson commented:
We haven't had any krb4 protocol code in MIT krb5 since release 1.7.  "v4" there just refers to a principal created using the v4 salt type, which is an empty salt.  The "password incorrect" error from kinit suggests that the string_to_key result in kadmin.local does not match the string_to_key result in kinit.  (This happens before we even rename the principal.)

You can re-run the test script with "--debug 7" and again with "--debug 8" to debug into the kadmin.local and kinit processes.  I'd suggest setting a breakpoint at krb5int_pbkdf2_hmac() and checking the arguments and behavior of PKCS5_PBKDF2_HMAC() with the empty salt.

(I tried reproducing this bug myself by checking out the PR branch and building krb5 against it, but for some reason it didn't fail for me.)

greghudson commented:
crypto/kdf/pbkdf2.c:173 sees that the value of salt is NULL and returns an error.  The krb5 side isn't doing error checking (our bad), so we use uninitialized memory for the result.

The documentation says "If the salt is NULL, then saltlen must be 0."  So I think the code is improperly checking the pointer alone.

mattcaswell commented:
Thanks @greghudson!

levitte commented:
I looked at the current master code, and the salt isn't checked at all there...  the salt *length* is checked down in the individual mixing functions (digest implementations), and they simply return 1 if that's zero.

Good catch, @greghudson!

levitte commented:
Another thing I notice that `PKCS5_PBKDF2_HMAC` does in master is to assign an empty string to the passed password if that happens to be NULL...

davidmakepeace commented:
> Another thing I notice that `PKCS5_PBKDF2_HMAC` does in master is to assign an empty string to the passed password if that happens to be NULL...

Yes, and that case is now handled in PKCS5_PBKDF2_HMAC() at crypto/evp/p5_crpt2.c:36 for compatibility.
I will add an additional check in PKCS5_PBKDF2_HMAC() for salt being NULL and saltlen being zero to ensure that the documented semantics are maintained.

davidmakepeace commented:
I added a fix for the PKCS5_PBKDF2_HMAC() salt == NULL issue and rebased to master.

paulidale commented:
It might be worthwhile to squash these down to a single commit with an all encompassing log message.

davidmakepeace commented:
> It might be worthwhile to squash these down to a single commit with an all encompassing log message.

Okay. I have just done that.

levitte commented:
Ok, I'm happy.  "nd reviewer needed.

FdaSilvaYY commented on crypto/kdf/kdf_util.c:
```diff
@@ -0,0 +1,73 @@
+
+#include <string.h>
+#include <stdarg.h>
+#include "internal/cryptlib.h"
```
nit: sort these include ? `"internal/` after `<openssl/`

davidmakepeace replied:
Fixed.

FdaSilvaYY commented on crypto/kdf/scrypt.c:
```diff
@@ -118,149 +129,377 @@ static int is_power_of_two(uint64_t value)
+static int kdf_scrypt_ctrl_uint32(EVP_KDF_IMPL *impl, int cmd,
+                                  const char *value)
+{
+    int int_value = atoi(value);
```
I think that `atoi` should be kick out of current codebase.
Should I open an issue ?

paulidale replied:
An issue is worth raising, if only to get us thinking about it.

FdaSilvaYY replied:
See #8216

FdaSilvaYY commented on crypto/kdf/scrypt.c:
```diff
@@ -118,149 +129,377 @@ static int is_power_of_two(uint64_t value)
+{
+    int i;
+    uint32_t x[16];
+    memcpy(x, inout, sizeof(x));
```
nit: blank line after declaration

davidmakepeace replied:
Fixed.

paulidale commented:
The build failures are unrelated to this.

davidmakepeace commented:
I just rebased, now that master has been fixed.
We should get a clean build now.

levitte commented:
Merged.

5a285addbf39f91d567f95f04b2b41764127950d Added new EVP/KDF API. Changed PKEY/KDF API to call the new API. Added wrappers for PKCS5_PBKDF2_HMAC() and EVP_PBE_scrypt() to call the new EVP KDF APIs. Documentation updated.

levitte commented:
This was great work, @davidmakepeace

paulidale commented:
219 days in coming.  A great effort and a very worthwhile contribution.  Thanks to all who assisted and contributed.

The various follow on KDFs can now be tapped onto this work.  @slontis has single step KDF, @romen has one too I believe.  No doubt there are others.  Let the floodgates open to the pressure of the 617th.

levitte commented:
[This search](https://github.com/openssl/openssl/pulls?utf8=%E2%9C%93&q=is%3Apr+is%3Aopen+kdf) shows all the KDF related PRs.  I'll have a quick run through, as I'm sure some can just be closed by now

tomato42 commented:
I must be missing something, but I don't see a way to run PBKDF2 using command line tools (see #4021), just the TLS KDF, HKDF and scrypt...

mattcaswell commented:
> I must be missing something, but I don't see a way to run PBKDF2 using command line tools (see #4021), just the TLS KDF, HKDF and scrypt...

It seems that PBKDF2 does not appear in crypto/evp/pkey_kdf.c like the others do. @davidmakepeace is this an accidental omission or was there a reason for that?

mattcaswell commented:
> I suppose only KDF algorithms that were previously implemented in the PKEY interface were ported via wrapping the KDF API. Anything newly implemented (PBKDF2, SSHKDF, ...) does not have this wrapping as the PKEY interface should be seen as legacy for kdfs?

Yes, perhaps that is the reason. That has the unfortunate side-effect that they're not available to pkeyutl. That might suggest that pkeyutl should be ported to use the new API - which is itself problematic since the whole structure of pkeyutl is focused around handling EVP_PKEY objects (hence the name of the application). Or alternatively we decide to retrofit the PKEY interface for these "new" KDFs.

levitte commented:
Didn't someone submit a PR to create a `openssl kdf' command?  I think I remember seeing that, but can't find it right now

davidmakepeace commented:
Yes, the PBKDF2 algorithm was not previously supported by the PKEY interface, which is now considered legacy with respect to KDF algorithms.
The intention was to eventually change everything that was using PKEY for KDFs to using the KDF API but there is no hurry due to the wrappers.

We could change the pkeyutl app to use the KDF API for KDFs. Alternatively a new kdf app could be added specifically for KDFs.

mattcaswell commented:
A KDF app seems like a good option to me.

slontis commented:
I had added a mac app, but not a kdf one.
I will submit a PR for this very soon.
Not sure that it is a good idea to use a utility like this to do anything other than do testing.
Deriving a key via a command-line tool seems a bit shady to me..

tomato42 commented:
there are two purposes, primary is to verify if a different implementation implements the KDF correctly[1], and secondary is for tools like https://testssl.sh/

so yes, diagnostics/debugging

 1 - including a situation where the KDF is implemented in an engine or pkcs11 module

slontis commented:
Added KDF app PR: https://github.com/openssl/openssl/pull/8762

slontis commented:
Added EVP_KDF PR: https://github.com/openssl/openssl/pull/8808
